File: client/js_src/src/app/game/state/players/movement/velocity.js
function validate_velocty(velocity, top_speed, frost) {
  if(frost != 0) {
    return validate_velocty_frost(velocity, top_speed, frost)
  } else if(velocity >= 0) {
    return Math.min(top_speed, velocity)
  } else {
    return Math.max(-top_speed, velocity)
  }
}

export function update_velocities(entity, ticks) {
  let result = Object.assign({}, entity);

  const acceleration = entity.acceleration;

  let vel_x_prev = entity.velocity_x;
  let vel_y_prev = entity.velocity_y;

  let vx = entity.velocity_x;
  let vy = entity.velocity_y;

  let accel_x = Math.min(entity.right, 1)*acceleration - Math.min(entity.left, 1)*acceleration
  let accel_y = Math.min(entity.down, 1)*acceleration -  Math.min(entity.up, 1) *acceleration

  vx += accel_x*ticks;
  vy += accel_y*ticks;

  let avg_vx = (entity.velocity_x + entity.velocity_x_prev)/2;
  let avg_vy = (entity.velocity_y + entity.velocity_y_prev)/2;

  let velocity_magnitude = Math.sqrt(Math.pow(avg_vy, 2) + Math.pow(avg_vx, 2));
  let friction_magnitude = velocity_magnitude * (entity.friction + entity.frost)
  if(velocity_magnitude != 0) {
    let friction_x = friction_magnitude * avg_vx/velocity_magnitude
    let friction_y = friction_magnitude * avg_vy/velocity_magnitude
    vx -=friction_x*ticks;
    vy -=friction_y*ticks;
  }


  result.velocity_x = vx;
  result.velocity_y = vy;
  result.velocity_x_prev = vel_x_prev;
  result.velocity_y_prev =  vel_y_prev;

  return result;
}

File: game_server/entities/player/lib/player/movement/velocity.ex
defmodule Player.Velocity do
  def update_velocity(player = %{velocity_x: vx, velocity_y: vy}, ticks) do
    {fdx, fdy} = get_friction(player, ticks)
    player |>
      update_velocity_x(vx, ticks) |>
      update_velocity_y(vy, ticks) |>
      Map.update!(:velocity_x, &(&1-fdx)) |>
      Map.update!(:velocity_y, &(&1-fdy)) |>
      Map.put(:velocity_x_prev, vx) |>
      Map.put(:velocity_y_prev, vy)
  end

  defp get_friction(%{velocity_x: vx, velocity_y: vy, velocity_x_prev: vxp, velocity_y_prev: vyp, friction: friction, frost: frost}, ticks) do
    avg_vx = (vx + vxp)/2
    avg_vy = (vy + vyp)/2
    velocity_magnitude = :math.sqrt(:math.pow(avg_vy, 2) + :math.pow(avg_vx, 2))

    if velocity_magnitude != 0 do
        friction_magnitude = velocity_magnitude * (friction + frost)
        friction_x = friction_magnitude * avg_vx/velocity_magnitude
        friction_y = friction_magnitude * avg_vy/velocity_magnitude
        {friction_x * ticks, friction_y * ticks}
      else
        {0, 0}
    end

  end

  defp update_velocity_x(player = %{left: accel_left, right: accel_right}, vx, ticks) do
    %{acceleration: acceleration} = player
    total_accel = min(1, accel_left)*acceleration - min(1, accel_right)*acceleration;
    Map.put(player, :velocity_x, vx - ticks*total_accel)
  end

  defp update_velocity_y(player = %{up: accel_up, down: accel_down}, vy, ticks) do
    %{acceleration: acceleration} = player
    total_accel = -min(1, accel_down)*acceleration + min(1, accel_up)*acceleration
    Map.put(player, :velocity_y, vy - ticks*total_accel)
  end

end
